phpBB/phpbb/image/image_cropper.php

```diff
@@ -0,0 +1,259 @@
+<?php
```

Comment:
Cannot this be done all in the front-end? If you can do all these (and probably more) from JS, I think it should be done there and just save the file.

Reply:
Well, it is possible to do it client side, most definitely.
However, that would enforce two things, one of which is fine and the other is debatable.
This would mean we would have to use `.toBlob()` on the image, and blobs can only be send with ajax request. So then we would have to ajaxify the form submission, which is fine.
However, the following:
> Uses the Browser's native [canvas.toBlob](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toBlob) API to do the compression work, which means it is **lossy compression**. For better image quality, you can upload the original image and the cropped data to a server and do the crop work on the server.

Reply:
Wait, you just send it as a hidden field. Why cannot we just set the front-end generated image as a value? Is it the problem that this library doesn't support that?

Reply:
No, the problem is that that will be at the cost of quality.
There are two possibilities:
1. Do the cropping client side and send a Blob, which is ready to go, but possibly losing quality.
2. Send the original image, with the cropping data, and do the cropping server side, to keep the quality.
I am currently using the 2nd option, so just sending the original image with the cropping data in a hidden field.

Reply:
You don't have to loose quality at all. Canvas can return whatever image compression quality you [want](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toDataURL).

Reply:
I'd also definitely opt for client side cropping